Budget App

ahzamir opened the pull request:
### In this pull request I have implemented all requirements below on my `budget-app`:

-Design
  - You should follow these [design guidelines](https://www.behance.net/gallery/19759151/Snapscan-iOs-design-and-branding?tracking_source=), including:
    - Colors.
    - Typography: font face, size, and weight.
    - Layout: composition and space between elements.
    
- Interactions
  - Splash screen

    - A simple page with the name of your app (yes, you need to choose one), and links to the sign-up and log-in pages.
Sign up and log in to pages

    - The user should be able to register in the app with their full name, email, and password (all mandatory).
    - The user can log into the app using email and password.
    - If the user is not logged in, they can't access pages that require the user to be logged in (all the pages described below).
- Home page (categories page)

  - When the user logs in, they are presented with the categories page.
  - For each category, the user can see their name, icon, and the total amount of all the transactions that belong to that category.
  - When the user clicks (or taps) on a category item, the application navigates to the transactions page for that category.
  - There is a button "add a new category" at the bottom that brings the user to the page to create a new category.
- Transactions page

  - For a given category, the list of transactions is presented, ordered by the most recent.
  - At the top of the page the user could see the total amount for the category (sum of all of the amounts of the transactions in that category).
  - There is a button "add a new transaction" at the bottom that brings the user to the page to create a new transaction.
  - When the user clicks on the "Back" button (<), the user navigates to the home page.
- "Add a new category" page

  - The user fills out a form to create a new category, indicating their name and icon (both mandatory).
  - The user clicks (or taps) the "Save" button to create the new category, and is taken to the home page on success.
  - When the user clicks on the "Back" button (<), the user navigates to the home page.
- "Add a new transaction" page

  - The user fills out a form to create a new transaction with:
    - name (mandatory)
    - amount (mandatory)
  - categories (mandatory at least one)
  - The user clicks (or taps) the "Save" button to create the new transaction, and is taken to the transactions page for that category.
  - When the user clicks on the "Back" button (<), the user navigates to the transactions page for that category.

- Testing requirements
  - Create unit and integration tests for all the most important components of your RoR application.
- Technical requirements
  - You should use Postgres as your database.
  - You should use devise for authentication.
  - You should validate all user input to make sure that anyone with bad intentions cannot compromise your app.
  - You can use a view template engine of your choice (.erb, .slim, .haml).
  - The project should be deployed and accessible online.

- Project documentation
  - Once you have finished the development of the project, you should record a video presenting the features of the project you built. It is a video with a maximum length of 5 minutes. The content of the video should include:

  - a description of the project.
  - a demo of the project features.
  - you should also highlight some interesting piece of code or something you built that you are very proud of.
For recording the video you can use tools like [Loom](https://www.loom.com/) that let you share a private link to the recording, and configure a shot that shows your computer screen and your face at the same time in a small picture.

ShadyShawkat commented on README.md:
```diff
@@ -1,74 +1,116 @@
+
+## Live Demo
+- [Video demo](https://www.loom.com/share/30f987d8de024de8b90b6ace0016a6d0)
+- [Budget_app - Comming soon]()
+
```
- [ ] Please make sure that your website is hosted and add the live demo link to your README.md file. As per [this](https://github.com/microverseinc/curriculum-rails/blob/main/capstone/rails_capstone.md#:~:text=Deploy%20the%20project%2C%20and%20test%20for%20final%20details.) requirement.

ahzamir replied:
@ShadyShawkat  I have an issue while deploying my app using Heroku. I am using WSL in windows.
This is the error I face while deploying.

![Screenshot 2022-09-22 171136](https://user-images.githubusercontent.com/96838030/191749845-7d6f0c8e-f84c-4c58-ada0-a3d4ca2e67b6.png)
![Screenshot 2022-09-22 171115](https://user-images.githubusercontent.com/96838030/191749877-349134cd-7649-4672-88cf-468a06d84e4a.png)

ahzamir replied:
@ShadyShawkat  Can you help me figure it out?

ShadyShawkat reviewed (changes requested):
## Hi @ahzamir, 

Good job so far! 👍🏻 
There are some issues that you still need to work on to go to the next project but you are almost there!

![Image](https://media.giphy.com/media/UdaSUYClr92cjWpeVx/giphy.gif)

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

ahzamir commented on spec/requests/categories_spec.rb:
```diff
@@ -0,0 +1,25 @@
+require 'rails_helper'
+
+RSpec.describe 'Categories', type: :request do
+  before(:each) do
+    @user = User.create!(name: 'Ahmad Zamir Yousufi', email: 'example@gmail.com', password: '123456')
+    @category = Category.create!(name: 'Test Category', icon: 'test-icon', user_id: @user.id)
+  end
+
+  describe 'GET /categories' do
+    it 'returns http success' do
+      get user_categories_path(@user)
+      expect(response).to have_http_status(:success)
+    end
+
+    it 'renders the index template' do
+      get user_categories_path(@user)
+      expect(response).to render_template('index')
+    end
+
+    it 'loads all of the categories' do
+      get user_categories_path(@user)
+      expect(assigns(:categories)).to eq([@category])
+    end
+  end
+end
```
@ShadyShawkat  Thank you very much, I solved it.

ahzamir commented:
## I have implemented all functionalities and required changes from the pervious reviewer.
- I have an issue while deploying my project. I have added my question on microverse discourse so I will deploy my project after I get an answer from them on how to solve the issue.

rotimiazeez reviewed (changes requested):
# Hi @ahzamir 👍🏽, 
## Your tests are now passing 💯  
###Great work on making the changes 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


## Suggested changes
_Check the comments under the review._
### If you are having a problem deploying to Heroku, I'd advise trying fly.io kindly read more about it [here](https://fly.io/docs/)

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

ahzamir commented:
@rotimiazeez  I have tried to deploy my project using fly.io but I couldn't can you please help me to deploy it?
- Can we please squeal a meeting?

AshakaE reviewed (changes requested):
Hi @ahzamir 🙋 , 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Required Changes ♻️
Kindly retry deploying the app to Heroku but on a new heroku link follow [heroku guide](https://devcenter.heroku.com/articles/getting-started-with-rails7)

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something
**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

ahzamir commented:
@AshakaE  Thank you very much for your kind review and feedback .
- The previous link was working fine, but I changed it as you requested. This is the new link for my deploying and everything is working fine. [Budget-app](https://zamir-budget-app.herokuapp.com/users/sign_in)

uwadonat reviewed (approved):
## Hi @ahzamir, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 
### Highlights
- Amazing design ✔️
- Very descriptive pull request message ✔️
![image](https://user-images.githubusercontent.com/65860678/191869427-f6b13651-ad51-4d1d-af6b-3de14e8a46db.gif)

# Status: Approved ✔️✔️✔️

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100%.
Please, remember to tag me @uwadonat in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
```

ahzamir commented:
@rotimiazeez @ShadyShawkat @uwadonat @AshakaE  Thank you very much for your kind review and feedback.
